File: src/usecases/workRecord.js
/* exported recordWorkStart, recordWorkEnd, recordDefect */

function recordWorkStart() {
  const { config } = getConfigSettingAndToken();
  
  // 自宅発送シートでアクティブな行のデータを取得
  const homeSheet = new HomeShipmentSheet(config.HOME_SHIPMENT_SHEET_NAME);
  const activeData = homeSheet.getActiveRowData();
  
  if (activeData.length === 0) {
    throw new Error('選択された行がありません');
  }
  
  // 作業記録シートに追加
  console.log(`作業記録シート名: ${config.WORK_RECORD_SHEET_NAME}`);
  const workRecordSheet = new WorkRecordSheet(config.WORK_RECORD_SHEET_NAME);
  const timestamp = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy/MM/dd HH:mm:ss');
  
  for (const row of activeData) {
    const asin = row.get("ASIN");
    const purchaseDate = row.get("購入日");
    let orderNumber = '';
    try { orderNumber = row.get("注文番号") || ''; } catch (e) {}
    
    if (!asin) {
      console.warn(`ASINが空の行をスキップしました`);
      continue;
    }
    
    workRecordSheet.appendRecord(asin, purchaseDate, "開始", timestamp, null, null, null, orderNumber);
  }
  
  console.log(`${activeData.length}件の作業記録（開始）を追加しました`);
}

function recordWorkEnd() {
  const { config } = getConfigSettingAndToken();
  
  // 自宅発送シートでアクティブな行のデータを取得
  const homeSheet = new HomeShipmentSheet(config.HOME_SHIPMENT_SHEET_NAME);
  const activeData = homeSheet.getActiveRowData();
  
  if (activeData.length === 0) {
    throw new Error('選択された行がありません');
  }
  
  // 作業記録シートに追加
  console.log(`作業記録シート名: ${config.WORK_RECORD_SHEET_NAME}`);
  const workRecordSheet = new WorkRecordSheet(config.WORK_RECORD_SHEET_NAME);
  const timestamp = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy/MM/dd HH:mm:ss');
  
  for (const row of activeData) {
    const asin = row.get("ASIN");
    const purchaseDate = row.get("購入日");
    let orderNumber = '';
    try { orderNumber = row.get("注文番号") || ''; } catch (e) {}
    
    if (!asin) {
      console.warn(`ASINが空の行をスキップしました`);
      continue;
    }
    
    workRecordSheet.appendRecord(asin, purchaseDate, "終了", timestamp, null, null, null, orderNumber);
  }
  
  console.log(`${activeData.length}件の作業記録（終了）を追加しました`);
}

function recordDefect() {
  const config = getEnvConfig();
  const ui = SpreadsheetApp.getUi();

  // 1. 自宅発送シートのS列から不良原因リストを読み込み
  const homeSheet = new HomeShipmentSheet(config.HOME_SHIPMENT_SHEET_NAME);
  const defectReasonList = homeSheet.getDefectReasonList();

  if (defectReasonList.length === 0) {
    ui.alert('エラー', '不良原因リストが見つかりません。自宅発送シートのS列に不良原因を入力してください。', ui.ButtonSet.OK);
    return;
  }

  // 2. 該当行の行番号を取得
  const activeData = homeSheet.getActiveRowData();
  if (activeData.length === 0) {
    ui.alert('エラー', '選択された行がありません。', ui.ButtonSet.OK);
    return;
  }

  // BaseRow(row) から直接、実シート上の行番号を取得する
  const rowNumbers = activeData.map(row => row.rowNumber).filter(rn => rn !== null && rn !== undefined && rn !== '');
  if (rowNumbers.length === 0) {
    ui.alert('エラー', '選択された行に有効な行番号がありません。', ui.ButtonSet.OK);
    return;
  }

  // 仕入管理シートの行番号とASIN/購入日/注文番号を取得
  const rowInfo = activeData.map(row => {
    let orderNumber = '';
    try { orderNumber = row.get("注文番号") || ''; } catch (e) {}
    return {
      homeRowNumber: row.rowNumber,
      purchaseRowNumber: row.get("行番号"),
      asin: row.get("ASIN"),
      purchaseDate: row.get("購入日"),
      orderNumber: orderNumber
    };
  }).filter(info => info.purchaseRowNumber);

  // 3. 不良数を入力
  const quantityResponse = ui.prompt('不良品登録', '不良数を入力してください:', ui.ButtonSet.OK_CANCEL);
  if (quantityResponse.getSelectedButton() !== ui.Button.OK) {
    return;
  }
  const quantity = parseInt(quantityResponse.getResponseText());
  if (isNaN(quantity) || quantity <= 0) {
    ui.alert('エラー', '有効な不良数を入力してください。', ui.ButtonSet.OK);
    return;
  }

  // 4. 原因を選択（番号で入力）
  const reasonList = defectReasonList.map((r, i) => `${i + 1}: ${r}`).join('\n');
  const reasonResponse = ui.prompt('不良品登録', `原因を番号で選択してください:\n${reasonList}`, ui.ButtonSet.OK_CANCEL);
  if (reasonResponse.getSelectedButton() !== ui.Button.OK) {
    return;
  }
  const reasonIndex = parseInt(reasonResponse.getResponseText()) - 1;
  if (isNaN(reasonIndex) || reasonIndex < 0 || reasonIndex >= defectReasonList.length) {
    ui.alert('エラー', '有効な番号を入力してください。', ui.ButtonSet.OK);
    return;
  }
  const selectedReason = defectReasonList[reasonIndex];

  // 5. コメント（任意）
  const commentResponse = ui.prompt('不良品登録', 'コメント（任意、空欄可）:', ui.ButtonSet.OK_CANCEL);
  if (commentResponse.getSelectedButton() !== ui.Button.OK) {
    return;
  }
  const comment = commentResponse.getResponseText() || null;

  // 6. 処理実行
  const purchaseRowNumbers = rowInfo.map(info => info.purchaseRowNumber);

  // 仕入管理シートの購入数を不良数分減らす
  const purchaseSheet = new PurchaseSheet(config.PURCHASE_SHEET_NAME);
  purchaseSheet.filter("行番号", purchaseRowNumbers);

  if (purchaseSheet.data.length === 0) {
    ui.alert('エラー', '仕入管理シートに対応する行が見つかりません。', ui.ButtonSet.OK);
    return;
  }

  const zeroQuantityRows = purchaseSheet.decreasePurchaseQuantity(quantity);

  // 購入数が0になった行を削除
  if (zeroQuantityRows.length > 0) {
    purchaseSheet.deleteRows(zeroQuantityRows);
  }

  // 作業記録に登録
  const workRecordSheet = new WorkRecordSheet(config.WORK_RECORD_SHEET_NAME);
  const timestamp = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy/MM/dd HH:mm:ss');

  for (const info of rowInfo) {
    const asin = info.asin;
    const purchaseDate = info.purchaseDate;
    const orderNumber = info.orderNumber || '';

    if (!asin) {
      console.warn(`ASINが空の行をスキップしました`);
      continue;
    }

    workRecordSheet.appendRecord(asin, purchaseDate, "不良", timestamp, quantity, selectedReason, comment, orderNumber);
  }

  console.log(`${rowInfo.length}件の不良品記録を追加しました`);
  ui.alert('完了', `不良品登録を完了しました。\n不良数: ${quantity}\n原因: ${selectedReason}`, ui.ButtonSet.OK);
}

File: python/src/usecases/work_record.py
from __future__ import annotations

import logging
from datetime import datetime

import click

from shared.config import AppConfig
from infrastructure.spreadsheet.base_sheets_repository import BaseSheetsRepository
from infrastructure.spreadsheet.home_shipment_sheet import HomeShipmentSheet
from infrastructure.spreadsheet.purchase_sheet import PurchaseSheet
from infrastructure.spreadsheet.work_record_sheet import WorkRecordSheet

logger = logging.getLogger(__name__)


def record_work_start(config: AppConfig, repo: BaseSheetsRepository, row_numbers: list[int]) -> None:
    home_sheet = HomeShipmentSheet(repo, config.sheet_id, config.home_shipment_sheet_name)
    home_sheet.get_rows_by_numbers(row_numbers)
    work_record = WorkRecordSheet(repo, config.sheet_id, config.work_record_sheet_name)
    timestamp = datetime.now().strftime("%Y/%m/%d %H:%M:%S")
    for row in home_sheet.data:
        asin = str(row.get("ASIN") or "").strip()
        purchase_date = str(row.get("購入日") or "").strip()
        try:
            order_number = str(row.get("注文番号") or "").strip()
        except Exception:
            order_number = ""
        if not asin:
            logger.warning("ASINが空の行をスキップしました")
            continue
        work_record.append_record(asin, purchase_date, "開始", timestamp, order_number=order_number)
    logger.info("%d件の作業記録（開始）を追加しました", len(home_sheet.data))


def record_work_end(config: AppConfig, repo: BaseSheetsRepository, row_numbers: list[int]) -> None:
    home_sheet = HomeShipmentSheet(repo, config.sheet_id, config.home_shipment_sheet_name)
    home_sheet.get_rows_by_numbers(row_numbers)
    work_record = WorkRecordSheet(repo, config.sheet_id, config.work_record_sheet_name)
    timestamp = datetime.now().strftime("%Y/%m/%d %H:%M:%S")
    for row in home_sheet.data:
        asin = str(row.get("ASIN") or "").strip()
        purchase_date = str(row.get("購入日") or "").strip()
        try:
            order_number = str(row.get("注文番号") or "").strip()
        except Exception:
            order_number = ""
        if not asin:
            logger.warning("ASINが空の行をスキップしました")
            continue
        work_record.append_record(asin, purchase_date, "終了", timestamp, order_number=order_number)
    logger.info("%d件の作業記録（終了）を追加しました", len(home_sheet.data))


def record_defect(config: AppConfig, repo: BaseSheetsRepository, row_numbers: list[int]) -> None:
    home_sheet = HomeShipmentSheet(repo, config.sheet_id, config.home_shipment_sheet_name)
    defect_reasons = home_sheet.get_defect_reason_list()
    if not defect_reasons:
        raise RuntimeError("不良原因リストが見つかりません")
    home_sheet.get_rows_by_numbers(row_numbers)
    if not home_sheet.data:
        raise RuntimeError("選択された行がありません")
    row_info = []
    for row in home_sheet.data:
        try:
            order_number = str(row.get("注文番号") or "").strip()
        except Exception:
            order_number = ""
        row_info.append({
            "purchase_row_number": str(row.get("行番号") or "").strip(),
            "asin": str(row.get("ASIN") or "").strip(),
            "purchase_date": str(row.get("購入日") or "").strip(),
            "order_number": order_number,
        })
    quantity = click.prompt("不良数を入力してください", type=int)
    if quantity <= 0:
        raise ValueError("有効な不良数を入力してください")
    reason_list = "\n".join(f"{i + 1}: {r}" for i, r in enumerate(defect_reasons))
    reason_index = click.prompt(f"原因を番号で選択してください:\n{reason_list}", type=int) - 1
    if reason_index < 0 or reason_index >= len(defect_reasons):
        raise ValueError("有効な番号を入力してください")
    selected_reason = defect_reasons[reason_index]
    comment = click.prompt("コメント（任意、空欄可）", default="", show_default=False) or None
    purchase_row_numbers = [info["purchase_row_number"] for info in row_info if info["purchase_row_number"]]
    purchase_sheet = PurchaseSheet(repo, config.sheet_id, config.purchase_sheet_name)
    purchase_sheet.filter("行番号", purchase_row_numbers)
    if not purchase_sheet.data:
        raise RuntimeError("仕入管理シートに対応する行が見つかりません")
    zero_rows = purchase_sheet.decrease_purchase_quantity(quantity)
    if zero_rows:
        purchase_sheet.delete_rows(zero_rows)
    work_record = WorkRecordSheet(repo, config.sheet_id, config.work_record_sheet_name)
    timestamp = datetime.now().strftime("%Y/%m/%d %H:%M:%S")
    for info in row_info:
        if not info["asin"]:
            continue
        work_record.append_record(
            info["asin"], info["purchase_date"], "不良", timestamp,
            quantity=quantity, reason=selected_reason, comment=comment,
            order_number=info["order_number"],
        )
    logger.info("%d件の不良品記録を追加しました", len(row_info))
    click.echo(f"不良品登録完了。不良数: {quantity}, 原因: {selected_reason}")
